### ISSUE TITLE
Name / id's are mixed & matched in provider / docs

### ISSUE THREAD

**Issue by u2mejc** [terraform-providers/terraform-provider-tfe#28]
**TL;DR:** Engineers often read the underlying API docs to understand the expected input that may otherwise be causing Terraform to fail cryptically. Renaming the human readable API objects in terraform HCL attributes creates vast confusion. Consider remapping all attribute names & IDs directly to their API counter parts, or choose a standard mapping scheme and stick with it.

**ISSUE:**
- `tfe_variable` Importing: When attempting to import a TFE variable, terraform will return successful, but then subsequently fail when attempting to read the value. Importing using the API's `id`, will return successful. The TFE variable API `id` isn't  visible in the UI, and doesn't match the scheme of the `workspace..id`, so it's not at all intuitive for a user.
```
# fails
terraform import tfe_variable.test <org>/<workspace name>/<var name>
# passes
terraform import tfe_variable.test <org>/<workspace name>/<var id from API>
```

- `tfe_variable` Creating: `tfe_variable workspace_id` perpetuates this confusion by requiring `workspace_id`, but this isn't the API's `id` for the workspace, instead this is expecting something like `.relationships.organization.data.id / .attributes.name`
```
# fails
workspace_id = "ws-abcdefghijkl"
# works
workspace_id = "<org_name>/<workspace_name>" # ie tfe_workspace..id
```

-  ID mismatch; `tfe_variable..id` vs` tfe_workspace..id`: _tfe_variable_ returns the API ID, where as the _tfe_workspace_ returns `.relationships.organization.data.id / .attributes.name` which seems to lack consistance, and isn't strait forward for the user.

- name and ID's are used interchangeably in documentation, leading to confusion that breaks external wrapper use if you actually want the name or ID (one may be missing today on some resources).
```
tfe_organization..id "The name of the organization."
vs
tfe_variable..id "The ID of the variable."
```
```
tfe_workspace..external_id = ws-abcdefghijkl
vs
tfe_workspace..id = <org>/<workspace_name>
```

**Recommendation**:
Introduce a breaking change now in version 0.3.0 of the TFE provider. Add `name` and `id` attributes for all resources, and map directly to the API's JSON objects with the same element names.

> name = name
> id = id

Update tfe_variable to track the human readable `<org name>/<workspace name>/<key name>` (instead of the hidden id value: `<org name>/<workspace name>/<key id>`)

**Comment by svanharmelen** [comment]
@u2mejc thank you for taking the time to open this issue and report your findings. Before responding to your specific examples and your recommendation, I would like to point out the difference between the internal resource ID of a Terraform resource (let's call this the `internal ID` for now) and the ID of that same resource in TFE (or any other SaaS service for that matter; let's call this the `SaaS ID`).

The `internal ID` of a Terraform resource is used the track the state of that resource over time. From the [docs](https://www.terraform.io/docs/extend/writing-custom-providers.html#implement-create): `This ID can be any string value, but should be a value that can be used to read the resource again.`

The most important part here is that the `internal ID` should be a value that enables Terraform to locate the resource so it can read its current settings and update the Terraform state.

In general this means that storing the `SaaS ID` of the resource as the `internal ID` is a good default as in most cases the SaaS APIs offer endpoints that can be used to retrieve the resource by it's `SaaS ID`.

But in some cases the APIs of the SaaS service do not offer the needed API endpoints to retrieve the required information using only the `SaaS ID`. In those cases it's perfectly fine to use a custom `internal ID` that enables you to retrieve the resource details.

This is exactly what we did in the TFE provider. Those resources that could be (fully) retrieved by their `SaaS ID`, use that same ID as the `internal ID`. But were the TFE API did not offer a means to retrieve all the resource details by `SaaS ID`, we used a custom `internal ID` instead.

In most use cases it shouldn't matter what is used as the `internal ID` as you don't need to set the ID as a user of the provider and the ID is there mainly for Terraform itself. The main exception for this would be when you import existing resources (which should be a one time custom action in most use cases). When importing resources you do have to know the exact `import ID` format that is expected by the resource. Again, in most cases this will be the `internal ID` but in some cases this can be a completely different custom `import ID`.

So with all that said let me try to reply on your specific examples:

```
# fails
terraform import tfe_variable.test <org>/<workspace name>/<var name>
# passes
terraform import tfe_variable.test <org>/<workspace name>/<var id from API>
```

This is correct and as far as I can tell this is properly documented [here](https://www.terraform.io/docs/providers/tfe/r/variable.html#import). In this case the `internal ID` of the variable is in fact set to the `SaaS ID` of the variable. But this import command required a custom `import ID`. The reason that this import command requires both the `internal ID` of a workspace (i.e. \<org\>/\<workspace name\>) and the `SaaS ID` (i.e. \<var-5rTwnSaRPogw6apb\>) of the variable, is that when retrieving the variable by `SaaS ID` the returned details do not contain the workspace that the variable belongs to. So in order to be able to import variables, we needed to get both these details.

```
# fails
workspace_id = "ws-abcdefghijkl"
# works
workspace_id = "<org_name>/<workspace_name>" # ie tfe_workspace..id
```

This is a good example of using a custom `internal ID` because the SaaS API does not offer endpoints to retrieve the resource by the `SaaS ID`. All the workspace related endpoints require the org and workspace names and do not accept the `SaaS ID` of the workspace. This is an inconsistency of the API that will hopefully be addressed in a future release of the TFE API, but for now the only solution is to use the custom `internal ID`.

I do understand the confusion in this case. In the TFE provider, fields that point to something like `foo_id` always refer to the `internal ID` of a `foo` resource. But I understand that this can also be given a string literal instead of a reference to another resource defined in your config. So I see how adding some details to the documentation about the expected ID format (just like what is done for the import commands) could make this more clear and easier to use.

```
tfe_organization..id "The name of the organization."
vs
tfe_variable..id "The ID of the variable."
```

This is again related to how the TFE API works. You can only get the organization by its name and all the API endpoints that expect a reference to the organization expect its name. So in that sense the ID to work with is the name and not the `SaaS ID` (which you cannot even retrieve from the API by the way).

```
tfe_workspace..external_id = ws-abcdefghijkl
vs
tfe_workspace..id = <org>/<workspace_name>
```

I think I already explained this one. The `external_id` here is the `SaaS ID` and the `id` is the `internal ID`. The related docs already describe this pretty well [here](https://www.terraform.io/docs/providers/tfe/r/workspace.html#attributes-reference). As for the exposed `id` field of a resource, this always exposing the `internal ID` of the resource.

Hopefully this info will give some more understanding about why things are the way they are. And I also hope that by now you understand that following your recommendations is not an option to solve this.

What we can do, is make it more explicit what kind (together with it's format) of ID is expected for which fields. I tried to add that info already to the import commands and a few fields, but there are probably a few other fields that could benefit for a more detailed description.

**Comment by u2mejc** [comment]
Thank you @svanharmelen for that explanation, I know it took some time and I really appreciate it. 👍 

As you commented, it sounds like there are some possible improvements for the API (luck be the owner of the API is also the owner of the TFE provider!). No need to waste time nit picking that with this API's current incarnation, it is possible to list all objects at a path, and then iterate for the resource(s) you want from that blob. Just wanted to share that the lack of unity is rather confusing and a little off putting at first. 😉

**Comment by svanharmelen** [comment]
Thanks @u2mejc!
